packages/web-react/src/components/Modal/useModal.tsx

```diff
@@ -0,0 +1,53 @@
+import { useState, useEffect, useCallback } from 'react';
+import { ClickEvent, SpiritModalProps } from '../../types';
+
+export interface useSpiritModalProps extends Pick<SpiritModalProps, 'isOpen' | 'onClose' | 'closeOnEscape'> {}
+export interface useSpiritModalReturn {
+  isOpen: boolean;
+  onClose: (event: ClickEvent | KeyboardEvent) => void;
+}
+
+export const useModal = (props: useSpiritModalProps) => {
+  const { isOpen, onClose, closeOnEscape, ...rest } = props;
+
+  const [modalOpen, setModalOpen] = useState(isOpen);
+
+  const closeHandler = useCallback(
+    (event: ClickEvent | KeyboardEvent) => {
+      setModalOpen(false);
+      if (onClose) onClose(event);
+    },
+    [onClose],
+  );
+
+  const userKeypressHandler = useCallback(
+    (event: KeyboardEvent) => {
```

Comment:
This handles the native `dialog` tag, eg. it captures the escape key down and fires `close` event.